File: prefix_sum/partitions_count/partitions_count.py

def count_max_partitions(nums: list[int], k: int) -> int:  # LeetCode Q.2025.
    total_sum, total_nums = sum(nums), len(nums)

    max_partition_ways, prefix_sum = 0, nums[0]
    for pivot in range(1, total_nums):
        if 2 * prefix_sum == total_sum:
            max_partition_ways += 1  # A natural pivot.
        prefix_sum += nums[pivot]

    right_side_diffs: dict[int, int] = dict()
    suffix_sum = 0
    for idx, num in enumerate(nums[::-1][:-1]):
        suffix_sum += num
        diff = total_sum - 2 * suffix_sum  # Diff = prefix sum - suffix sum.
        if diff not in right_side_diffs.keys():
            right_side_diffs[diff] = 0
        right_side_diffs[diff] += 1

    left_side_diffs: dict[int, int] = dict()
    prefix_sum = 0
    for idx, num in enumerate(nums):
        partition_ways = 0

        right_change = num - k
        if right_change in right_side_diffs.keys():
            partition_ways += right_side_diffs[right_change]

        left_change = k - num
        if left_change in left_side_diffs.keys():
            partition_ways += left_side_diffs[left_change]

        if partition_ways > max_partition_ways:
            max_partition_ways = partition_ways

        prefix_sum += num
        if idx > 0:  # Suffix doesn't cover num at 0th idx.
            suffix_sum -= num

        diff = prefix_sum - suffix_sum  # Diff = prefix sum - suffix sum.
        if diff not in left_side_diffs.keys():
            left_side_diffs[diff] = 0
        left_side_diffs[diff] += 1

        if diff in right_side_diffs.keys():
            right_side_diffs[diff] -= 1

    return max_partition_ways

File: prefix_sum/partitions_count/partitions_count.cpp
#include <vector>
#include <unordered_map>
using namespace std;

int countMaxPartitions(vector<int> &nums, int k)
{ // LeetCode Q.2025.
    // Sums related variables must use long long to prevent overflow.
    // Counts and indices related variables can use int.
    long long totalSum = 0;
    for (auto num : nums)
        totalSum += num;

    long long maxPartitionWays = 0, prefixSum = nums[0];

    for (int pivot = 1; pivot < nums.size(); pivot++)
    {
        if (2 * prefixSum == totalSum)
            maxPartitionWays += 1; // A natural pivot.

        prefixSum += nums[pivot];
    }

    unordered_map<long long, int> rightSideDiffs;
    long long suffixSum = 0;

    for (int idx = nums.size() - 1; idx >= 1; idx--)
    {
        suffixSum += nums[idx];
        long long diff = totalSum - 2 * suffixSum; // Diff = prefix sum - suffix sum.
        rightSideDiffs[diff]++;
    }

    unordered_map<long long, int> leftSideDiffs;
    prefixSum = 0;

    for (int idx = 0; idx < nums.size(); idx++)
    {
        long long partitionWays = 0;

        long long rightChange = nums[idx] - k;
        if (rightSideDiffs.find(rightChange) != rightSideDiffs.end())
            partitionWays += rightSideDiffs[rightChange];

        long long leftChange = k - nums[idx];
        if (leftSideDiffs.find(leftChange) != leftSideDiffs.end())
            partitionWays += leftSideDiffs[leftChange];

        if (partitionWays > maxPartitionWays)
            maxPartitionWays = partitionWays;

        prefixSum += nums[idx];
        if (idx > 0) // Suffix doesn't cover num at 0th idx.
            suffixSum -= nums[idx];

        long long diff = prefixSum - suffixSum;
        leftSideDiffs[diff]++;
        rightSideDiffs[diff]--;
    }

    return maxPartitionWays;
}
